src/hexkit/providers/akafka/provider/eventsub.py

```diff
@@ -177,76 +210,210 @@ async def construct(
                 event_value.decode("ascii")
             ),
         )
+
         try:
             await consumer.start()
-            yield cls(consumer=consumer, translator=translator)
+            yield cls(
+                consumer=consumer,
+                translator=translator,
+                publisher=publisher,
+                config=config,
+            )
         finally:
             await consumer.stop()
 
     def __init__(
-        self, *, consumer: KafkaConsumerCompatible, translator: EventSubscriberProtocol
+        self,
+        *,
+        consumer: KafkaConsumerCompatible,
+        translator: EventSubscriberProtocol,
+        config: KafkaConfig,
+        publisher: Optional[EventPublisherProtocol] = None,
     ):
         """Please do not call directly! Should be called by the `construct` method.
         Args:
             consumer:
-                hands over a started AIOKafkaProducer.
+                hands over a started AIOKafkaConsumer.
             translator (EventSubscriberProtocol):
                 The translator that translates between the protocol (mentioned in the
                 type annotation) and an application-specific port
                 (according to the triple hexagonal architecture).
+            publisher:
+                running instance of publishing provider that implements the
+                EventPublisherProtocol, such as KafkaEventPublisher. Can be None if
+                not using the dead letter queue.
+            config:
+                the KafkaRetryConfig instance containing dlq topics and retry allowance.
         """
         self._consumer = consumer
         self._translator = translator
         self._types_whitelist = translator.types_of_interest
-
-    @staticmethod
-    def _get_event_label(event: ConsumerEvent) -> str:
-        """Get a label that identifies an event."""
-        return (
-            f"{event.topic} - {event.partition} - {event.key} - {event.offset}"
-            + " (topic-partition-offset)"
+        self._publisher = publisher
+        self._dlq_topic = config.kafka_dlq_topic
+        self._retry_topic = config.kafka_retry_topic
+        self._max_retries = config.kafka_max_retries
+        self._enable_dlq = config.kafka_enable_dlq
+
+    def _get_original_topic(self, event: ConsumerEvent) -> str:
+        """Get the topic to use -- either the given topic or the one from `_original_topic`."""
+        topic = event.topic
+        if topic == self._retry_topic:
+            topic = str(event.value.get(ORIGINAL_TOPIC_FIELD, ""))
+            logging.info(
+                "Received previously failed event from topic '%s' for retry.", topic
+            )
+        return topic
+
+    async def _publish_to_dlq(self, *, event: ExtractedEventInfo):
+        """Publish the event to the DLQ topic."""
+        dlq_payload = {**event.payload, ORIGINAL_TOPIC_FIELD: event.topic}
```

Comment:
Wouldn't it make more sense to put the original topic into the header instead of the payload? I believe it is intended to be used for this kind of metadata.

Reply:
That's what I originally wanted to do but felt like the protocol was too holy to change

Reply:
The payload is even holier 😇

Reply:
I ran with that and modified the event pub protocol in [ee00b5e](https://github.com/ghga-de/hexkit/pull/123/commits/ee00b5e224ef53790692f82b54f41e0d1b1d77ba). The consequence is that this is a breaking change, so the branch was rebased to v4 and the PR retargeted.